rust/trezor-client/src/transport/udp.rs

```diff
@@ -0,0 +1,153 @@
+use super::{
+    error::Error,
+    protocol::{Link, Protocol, ProtocolV1},
+    AvailableDeviceTransport, ProtoMessage, Transport,
+};
+use crate::{AvailableDevice, Model};
+use std::{fmt, net::UdpSocket, result::Result, time::Duration};
+
+// A collection of constants related to the Emulator Ports.
+mod constants {
+    pub const DEFAULT_HOST: &str = "127.0.0.1";
+    pub const DEFAULT_PORT: &str = "21324";
+    pub const DEFAULT_DEBUG_PORT: &str = "21325";
+    pub const LOCAL_LISTENER: &str = "127.0.0.1:34254";
+}
+
+use constants::{DEFAULT_DEBUG_PORT, DEFAULT_HOST, DEFAULT_PORT, LOCAL_LISTENER};
+
+/// The chunk size for the serial protocol.
+const CHUNK_SIZE: usize = 64;
+
+const READ_TIMEOUT_MS: u64 = 100000;
+const WRITE_TIMEOUT_MS: u64 = 100000;
+
+/// An available transport for connecting with a device.
+#[derive(Debug)]
+pub struct AvailableUdpTransport {
+    pub host: String,
+    pub port: String,
+}
+
+impl fmt::Display for AvailableUdpTransport {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "udp:{}:{}", self.host, self.port)
+    }
+}
+
+/// An actual serial HID USB link to a device over which bytes can be sent.
+struct UdpLink {
+    pub socket: UdpSocket,
+    pub device: (String, String),
+}
+// No need to implement drop as every member is owned
+
+impl Link for UdpLink {
+    fn write_chunk(&mut self, chunk: Vec<u8>) -> Result<(), Error> {
+        debug_assert_eq!(CHUNK_SIZE, chunk.len());
+        let timeout = Duration::from_millis(WRITE_TIMEOUT_MS);
+        self.socket.set_write_timeout(Some(timeout))?;
+        if let Err(e) = self.socket.send(&chunk) {
+            return Err(e.into());
+        }
+        Ok(())
+    }
+
+    fn read_chunk(&mut self) -> Result<Vec<u8>, Error> {
+        let mut chunk = vec![0; CHUNK_SIZE];
+        let timeout = Duration::from_millis(READ_TIMEOUT_MS);
+        self.socket.set_read_timeout(Some(timeout))?;
+
+        let n = self.socket.recv(&mut chunk)?;
+        if n == CHUNK_SIZE {
+            Ok(chunk)
+        } else {
+            Err(Error::DeviceReadTimeout)
+        }
+    }
+}
+
+impl UdpLink {
+    pub fn open(path: &str) -> Result<UdpLink, Error> {
+        let mut parts = path.split(':');
+        let link = Self {
+            socket: UdpSocket::bind(LOCAL_LISTENER)?,
```

Comment:
why is this a fixed address instead of auto assigned?

Reply:
Fixed by changing `LOCAL_LISTENER` to 127.0.0.1:0, where the port will be auto-assigned.